### PR TITLE
No elmi

### PR DESCRIPTION
**Pull request by harrysarson**
Applying the work done in <https://github.com/rtfeldman/node-test-runner/pull/442> to the rust port of the test runner. Credit for the should go to @lydell as this is a port of their work.

Interested to hear feedback on:

- [ ] Error handling needs some thought.
- [x] Git submodules are pretty horrible. If we like this approach I will try get tree-sitter-elm published to crates.io so we can depend on it in the normal way.
- [ ] I need to have another look over <https://github.com/rtfeldman/node-test-runner/pull/442> for things I missed when porting.

### CODE REVIEW

File: templates/Runner.elm [hunk 1]
```diff
@@ -1,18 +1,56 @@
...
+If you rename or change this function you also need to update the regex that looks for it.
+-}
+check : a -> Maybe Test
+check =
+    checkHelperReplaceMe___
+
+
+checkHelperReplaceMe___ : a -> b
+checkHelperReplaceMe___ _ =
+    Debug.todo "The regex for replacing this Debug.todo with some real code must have failed since you see this message!\n\nPlease report this bug: https://github.com/rtfeldman/node-test-runner/issues/new\n"
```
Comment (lydell) [inline templates/Runner.elm:33]: Should link to https://github.com/mpizenberg/elm-test-rs/issues/new ?

File: src/run.rs [hunk 2]
```diff
@@ -325,3 +346,75 @@ fn create_templated<P: AsRef<Path>>(template: P, output: P, replacements: Vec<(S
...
+fn add_kernel_test_checking(elm_js: &str) -> String {
+    lazy_static::lazy_static! {
+
+        /// For older versions of elm-explorations/test we need to list every single
+        /// variant of the `Test` type. To avoid having to update this regex if a new
+        /// variant is added, newer versions of elm-explorations/test have prefixed all
+        /// variants with `ElmTestVariant__` so we can match just on that.
+        /// TODO(harry): ask Lydell if the \s*\$:\s*(['"])\1\2 bit is important.
+        /// I had to remove this from the end because the regex crate does not
+        /// support them.
```
Comment (lydell) [inline src/run.rs:359]: `(['"])\1\2` works like this:

- `(['"])` matches either a single quote or a double quote and remembers which one. `\2` then matches the same type of quote again. The Elm compiler always uses the same type of quote as far as I know (though I don’t remember which one), so this isn’t strictly needed. I just tried to future-proof.

- `\1` matches whatever was matched by `(ElmTestVariant__\w+|UnitTest|FuzzTest|Labeled|Skipped|Only|Batch)` earlier. The function name and the tag go together.

Your version looks good, though. I’m not sure why I bothered checking more. Maybe we should use your simpler version in node-test-runner, too.

File: src/parser.rs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        );
+        helper(
+            "port module Main.Pain exposing (int, Int, {- -}test, Test)",
+            &Some(vec!["int", "test"]),
+        );
+        helper(
+            "port module Main.Pain exposing (int, Int, -- comment
+    test, Test)",
+            &Some(vec!["int", "test"]),
+        );
```
Comment (lydell) [inline src/parser.rs:154]: Here’s a much more complicated comments test for inspiration: https://github.com/rtfeldman/node-test-runner/blob/ec4ec7312d2a415d9c22acb620f22aaa438192ed/tests/Parser.js#L32-L47

**Review by lydell — COMMENTED** [review]
Exciting! 🎉

**Comment by mpizenberg** [comment]
Thanks a lot for this! If you don't mind, I'd prefer keeping my focus on finishing my current work on pubgrub-rs before switching back to this. Didn't know you'd port that so fast ^^. I'll probably not answer for a couple of weeks but feel free to continue discussing things here and making progress on that with @lydell until I join the discussion later :)

**Comment by harrysarson** [comment]
> Didn't know you'd port that so fast

It turned out to be quite straightforward! I guess the hard bit is speccing out the design space, "code being the easy part" after all.

**Comment by harrysarson** [comment]
> Exciting! :tada:

Thanks for the feedback :) I will let this sit for a couple weeks before starting the polishing just to give @mpizenberg some time.

**Comment by mpizenberg** [comment]
Just a small update to let you know I'll soon be able to review this, in the coming week very likely

**Comment by harrysarson** [comment]
I have got the tree sitter ball rolling here https://github.com/Razzeee/tree-sitter-elm/pull/57

File: src/parser.rs (hunk 3 above)
Comment (harrysarson) [reply src/parser.rs:175]: I ported those tests to rust. Thanks!

File: src/run.rs (hunk 2 above)
Comment (harrysarson) [reply src/run.rs]: Thanks, this makes a lot of sense. The rust regex crate is really nice. Apparently they do not support the `(['"])\1\2` syntax:

> Its syntax is similar to Perl-style regular expressions, but **lacks a few features like look around and backreferences**. In exchange, all searches execute in linear time with respect to the size of the regular expression and search text.

That is a trade off I am happy they made.

**Comment by mpizenberg** [comment]
I've had a very quick look at the shape of the PR (not the logic yet). I feel like there are quite a few things that add noise because they are not directly related to the "no elmi" implementation. I'll try to extract those first in another PR to have a cleaner diff. "Make the change easy" as would say Dillon.

I've another high-level question. The PR adds a build step using the "cc" crate. I suppose that's because of tree sitter. Is it going to be an issue for portability and to generate a statically compiled executable, preferably based on musl?

Regarding the logic of the PR, would you mind do a summary of the necessary changes to what's existing and of what's new?

**Comment by harrysarson** [comment]
> I'll try to extract those first in another PR to have a cleaner diff. 

I can have a go at that, mainly changes to `src/run.rs` I think. Saying that there are not many unrelated changes, this parser approach requires slightly different input information and produces slightly different output information so there are needed changes to the run code to accommodate.

> I've another high-level question. The PR adds a build step using the "cc" crate. I suppose that's because of tree sitter. Is it going to be an issue for portability and to generate a statically compiled executable, preferably based on musl?

I can and will remove the dev dep on `cc` (it is a left over from the first draft of this PR when tree-sitter-elm had not yet been published to crates.io. There will be no custom build step in this repository. 

Saying that, there will still be a build step compiling the tree-sitter c code (just managed by the tree-sitter-elm crate rather than by us). The libraries created by the build step are statically linked into the executable and will use the configured compiler. Having a build step is 100% compatible with static musl binaries.

> Regarding the logic of the PR, would you mind do a summary of the necessary changes to what's existing and of what's new?

Certainly, watch this space...

**Comment by mpizenberg** [comment]
I've updated master and merged changes here to remove things unrelated to the PR. I'll take some time in the coming days to remind myself how elm-test-rs works (it has been some time ^^) to better understand what that PR brings.

**Comment by mpizenberg** [comment]
I've started by looking at the logic to get modules names. I'm wondering, would it be possible to extract modules names inside `parser::all_tests` such that `TestModule` has a `name` field that would be extracted by tree sitter? That would avoid all the code related to `get_module_name`.

**Comment by lydell** [comment]
> extract modules names

Do you mean extracting this part?

```
module Some.Name exposing
       ^^^^^^^^^
```

If so, just as a heads up: In node-test-runner we don’t use that module name. We just skip over it when parsing. Instead, we calculate the module name from the file path and source directories. This is because you can’t trust the module name – it can be wrong. In node-test-runner we don’t want to generate an `import Some.Name` where that points to no existing file – that gave bad error messages. Instead we ensure our generated imports are correct so the Elm compiler can report the `module` line being wrong in the user’s file. Sorry if this isn’t at all what you’re talking about.

https://github.com/rtfeldman/node-test-runner/blob/dafa12e58043915bdd8fcd7d2231ccff511a7827/lib/Runner.js#L39-L45

**Comment by mpizenberg** [comment]
I see. I thought there was no point double checking that this line is correct since the elm compiler also does it. In elm-test-rs we do this this just before parsing all tests:
```rust
    eprintln!("Compiling all test files ...");
    compile(
        &tests_root,       // current_dir
        &options.compiler, // compiler
        "/dev/null",       // output
        &module_paths,     // src
    );
```
where this basically calls `elm make [module_paths]... output=/dev/null`.
At that point I thought the elm compiler already checks that all tests file are valid including their `import` statements, and we can safely use those lines.

**Comment by mpizenberg** [comment]
Actually, that compilation step was needed before in order to get the `.elmi` files. It seems though that this compilation is not needed anymore, so no safe net anymore if I remove it.

**Comment by mpizenberg** [comment]
So if I'm getting it right, the function `add_kernel_test_checking` patches `Test` variants definitions to detect any user definition of type `Test` at runtime when initializing a worker for `Runner.elm`. It works thanks to the fact that any test is defined with one the variants of the base `Test` type, and because the elm compiler does not do optimizations such as inlining that could sidestep the variants patch.

If that is right, I've only one thing left to review, the `parse::all_tests` function. I have not read it yet, but am right to assume that given `add_kernel_test_checking` checks if something is a test, `parse::all_tests` would actually return all top level definitions, and not all tests?

**Comment by lydell** [comment]
> `parse::all_tests` would actually return all exposed top level definitions, and not all tests?

I don’t know about this Rust version, but for node-test-runner this is _almost_ correct. In node-test-runner, the function is called `extractExposedPossiblyTests`. It extracts everything that _could_ be a test, so not `Uppercase` exposed stuff, and not `f x =` (when `exposing (..)`, because a function can never be a `Test`, and this means we can simplify the parser by only looking for names followed by an equals sign basically).

**Comment by mpizenberg** [comment]
> It works thanks to the fact that any test is defined with one the variants of the base Test type

Just realizing, this would not detect a "test in the making" that would be temporarily defined by `Debug.todo "write this test"`. But I suppose that's a very niche case. Do you deal with that in node-test-runner?

**Comment by mpizenberg** [comment]
> Just realizing, this would not detect a "test in the making" that would be temporarily defined by Debug.todo "write this test".

Just just realizing that this would be a constant definition with a `Debug.todo` and those make the program crash on start anyway so we don't need to care about this

**Comment by mpizenberg** [comment]
Is tree-sitter a parser generator? I mean, the tree-sitter-elm repo seems to mainly contain JS code, and yet it seems that the [build step for the rust package](https://github.com/Razzeee/tree-sitter-elm/blob/master/rust/build.rs) only build two files, `parser.c` and `scanner.cc`. So I guess those are generated by the logic in the JS code?

**Comment by lydell** [comment]
I don’t know much about tree-sitter but I managed to find this:

https://tree-sitter.github.io/tree-sitter/

> Tree-sitter is a parser generator tool and an incremental parsing library. It can build a concrete syntax tree for a source file and efficiently update the syntax tree as the source file is edited.

https://tree-sitter.github.io/tree-sitter/creating-parsers

> In order to develop a Tree-sitter parser, there are two dependencies that you need to install:
>
> - Node.js - Tree-sitter grammars are written in JavaScript, and Tree-sitter uses Node.js to interpret JavaScript files. It requires the node command to be in one of the directories in your PATH. You’ll need Node.js version 6.0 or greater.
> - A C Compiler - Tree-sitter creates parsers that are written in C. In order to run and test these parsers with the tree-sitter parse or tree-sitter test commands, you must have a C/C++ compiler installed. Tree-sitter will try to look for these compilers in the standard places for each platform.

File: src/parser.rs [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+/// Find all possible tests (all values) in `test_files`.
+pub fn all_tests(
+    test_sources: impl IntoIterator<Item = (impl AsRef<Path>, impl AsRef<str>)>,
+) -> Result<Vec<TestModule>, String> {
+    test_sources
+        .into_iter()
+        .map(|(file_path, source)| {
+            let tree = {
+                let mut parser = tree_sitter::Parser::new();
+                let language = tree_sitter_elm::language();
```
Comment (mpizenberg) [inline src/parser.rs:111]: Could it make sense to have a much simpler parser able to parse modules headers. If that parser discovers `exposing(..)` we switch to full file elm parsing.

**Comment by mpizenberg** [comment]
It was quite some fun playing around the tree-sitter parser so I wanted to try its query system. I've rewritten `get_explicit_exposed_values` in a way that uses tree-sitter queries in [that last commit](https://github.com/mpizenberg/elm-test-rs/pull/15/commits/1bad6ac82084c64c84965647bdc5a5f6f17c795b). What do you think?

File: src/parser.rs [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    next_sibling(&mut cursor)?;
+    check_kind(cursor.node(), "left_parenthesis")?;
+    next_sibling(&mut cursor)?;
+
+    let ret = if cursor.node().kind() == "double_dot" {
+        next_sibling(&mut cursor)?;
+        None
+    } else {
+        let mut v = vec![];
+        while match cursor.node().kind() {
```
[the remark this answers is not in the document]
Comment (mpizenberg) [reply src/parser.rs:99]: Solved by query matches

**Comment by mpizenberg** [comment]
I've moved all the code related to the query approach at the end of the file: https://github.com/mpizenberg/elm-test-rs/blob/77b762e38d35b3f30bff5bc5a541311f845dff06/src/parser.rs#L347

One advantage I can see is that we could get rid of the `ChildCursor` struct, the `check_kind`, `skip_comments`, `child`, `next_sibling` helper functions, and the `ExplicitExposedValuesError` type.

**Comment by mpizenberg** [comment]
I cleaned up by removing the tree-walking approach so that only the code for the query approach is left. It looks quite nice in my opinion but we can revert the commit if there are things you don't like.

**Comment by harrysarson** [comment]
Hey! This is looking great! To apologise the product of free time and motivation that powers open source is looking very low for me at the moment and with Christmas coming it might stay that way to a while.

I hadn't seen the query system, it looks much better than what I had. I always felt that that must be a better API but I couldn't find it, looks like it is queries!

I like the new changes a lot :) feel free to run with this, I will try to chip in when I can :)

**Comment by mpizenberg** [comment]
I feel you, I have those time too ^^, thanks for the feedback @harrysarson and of course take the time you need, no pressure.

File: templates/Runner.elm [hunk 6]
```diff
@@ -36,24 +36,10 @@ Please report this bug: https://github.com/mpizenberg/elm-test-rs/issues/new
...
-    testsForModule
-        |> List.filterMap
-            (\{ module_, maybeTests } ->
-                case maybeTests |> List.filterMap (\x -> x) of
-                    [] ->
-                        Nothing
-
-                    tests ->
-                        Just (Test.describe module_ tests)
-            )
```
Comment (lydell) [inline templates/Runner.elm]: Did you really mean to remove all of this? I understand changing `\x -> x` to `identity`, but we still want `Test.describe module_` etc?

Comment (mpizenberg) [reply templates/Runner.elm]: Last iteration is
```elm
main : Program Flags Model Msg
main =
    [ {{ potential_tests }} ]
        |> List.filterMap check
        |> Test.concat
        |> ...
```
Do we really want the `Test.describe module ...` ?

Comment (lydell) [reply templates/Runner.elm]: I think `Test.describe module` makes it so you can have `test "foo"` in two files without getting a “duplicate test name” error.

Comment (mpizenberg) [reply templates/Runner.elm]: For people putting all their tests in `testst/Tests.elm` it is not that useful. For people putting different tests in different files, it is more or less the same work doing
```elm
module ... exposing (a, b, c)
```
or doing
```elm
module ... exposing (tests)
tests = describe "whatever" [a, b, c]
```

I've always wondered where that first line in an error report was coming from. At least like this it's explicit. What do you think?

Comment (mpizenberg) [reply templates/Runner.elm]: > I think Test.describe module makes it so you can have test "foo" in two files without getting a “duplicate test name” error.

Oh I see ... hum I don't know how I feel about that.

Comment (lydell) [reply templates/Runner.elm]: Not sure either. Are people supposed to be able to switch to elm-test-rs without having to change any Elm code?

Comment (lydell) [reply templates/Runner.elm]: It’s also visible when tests fail. If you have top-level tests in a file (not inside any `describe`), you get a clue which file the failing tests come from.

Comment (mpizenberg) [reply templates/Runner.elm]: > Are people supposed to be able to switch to elm-test-rs without having to change any Elm code?

I'll add a "Differences with elm-test" section in the readme. That's preferable in my opinion. Being a complete rewrite, I'd not be surprised if there are some other unexpected difference anyway.

**Comment by mpizenberg** [comment]
@harrysarson @lydell Thanks a lot for this contribution! I'm about to merge this. Do you mind if I add the two of you in the [`authors` field of the `Cargo.toml` file](https://github.com/harrysarson/elm-test-rs/blob/no-elmi/Cargo.toml#L4)? And if that's ok, which email should I include?

**Comment by lydell** [comment]
Thanks! At this point, I’d rather not be listed as an author, since I have 0 commits and don’t know Rust. A few bits of comments and Elm and maybe test cases might be copied from stuff I wrote, but I don’t mind. I’m just happy to have been able to help!

**Comment by harrysarson** [comment]
I wouldn't say no! Thank you that is very kind.
harry.sarson@hotmail.co.uk :)

**Comment by mpizenberg** [comment]
Awesome, I'll add a commit on master directly :+1:

**Comment by mpizenberg** [comment]
Some notes from recent tests. The package elm-geometry is one such package that exhibits a duplicate test names error with elm-test-rs. Here is the result of me searching for one such conflicting name:
```sh
$> rg "Every point on an arc is within its bounding box" tests/
tests/Tests/EllipticalArc2d.elm
159:        "Every point on an arc is within its bounding box"

tests/Tests/Arc3d.elm
118:        "Every point on an arc is within its bounding box"

tests/Tests/Arc2d.elm
183:        "Every point on an arc is within its bounding box"

tests/Tests/EllipticalArc3d.elm
87:        "Every point on an arc is within its bounding box"
```

Also in my last few commits when I extracted `check` out into the line `|> List.filterMap check` I introduced a bug. Indeed the list above is not homogeneous before applying the `check` function. So I've now restored the behavoir of inlining the `check` for the list items and filtering with `List.filterMap identity`.

**Pull request merged**